lightbeam/validate.py

```diff
@@ -108,33 +292,136 @@ def load_local_descriptors(self):
         return ""
 
+    # Validates descriptor values for a single payload (returns an error message or empty string)
+    def has_invalid_references(self, payload, path=""):
+        for k in payload.keys():
+            if isinstance(payload[k], dict) and not k.endswith("Reference"):
+                value = self.has_invalid_references(payload[k], path+("." if path!="" else "")+k)
+                if value!="": return value
+            elif isinstance(payload[k], list):
+                for i in range(0, len(payload[k])):
+                    value = self.has_invalid_references(payload[k][i], path+("." if path!="" else "")+k+"["+str(i)+"]")
+                    if value!="": return value
+            elif isinstance(payload[k], dict) and k.endswith("Reference"):
+                is_valid_reference = False
+                original_endpoint = util.pluralize_endpoint(k.replace("Reference",""))
+
+                # this deals with the fact that an educationOrganizationReference may be to a school, LEA, etc.:
+                endpoints_to_check = self.EDFI_GENERICS_TO_RESOURCES_MAPPING.get(original_endpoint, [original_endpoint])
+                for endpoint in endpoints_to_check:
+                    # check if it's a local reference:
+                    if endpoint not in self.local_reference_cache.keys(): break
+                    # construct cache_key for reference
+                    cache_key = self.get_cache_key(payload[k])
+                    if cache_key in self.local_reference_cache[endpoint]:
+                        is_valid_reference = True
+                        break
+                if not is_valid_reference: # not found in local data...
+                    for endpoint in endpoints_to_check:
+                        # check if it's a remote reference:
+                        params = payload[k].copy()
+                        if "link" in params.keys(): del params["link"]
+                        value = self.remote_reference_exists(endpoint, params)
+                        if value:
+                            is_valid_reference = True
+                            break
+                    if not is_valid_reference:
+                        return f"payload contains an invalid {k} " + (" (at "+path+"): " if path!="" else ": ") + json.dumps(params)
+        return ""
+
     # Tells you if a specified descriptor value is valid or not
     def is_valid_descriptor_value(self, namespace, codeValue):
         for row in self.lightbeam.api.descriptor_values:
             if row[1]==namespace and row[2]==codeValue:
                 return True
         return False
+
+    @staticmethod
+    def get_cache_key(payload):
+        cache_key = ''
+        for k in payload.keys():
+            cache_key += f"{payload[k]}~~~"
+        return cache_key
     
+    def remote_reference_exists(self, endpoint, params):
+        # check cache:
+        if endpoint=='students' and 'studentUniqueId' in params.keys(): return True
```

Comment:
Just to make absolutely sure we don't forget, flagging this line that you discovered while debugging